backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/PluginsSynchronizer.java

```diff
@@ -58,15 +58,25 @@ public PluginSynchronizationSummary synchronize(ServerApi serverApi, boolean sup
 
     var storedPluginsByKey = storage.plugins().getStoredPluginsByKey();
     var serverPlugins = serverApi.plugins().getInstalled(cancelMonitor);
-    var pluginsToDownload = serverPlugins.stream()
-      .filter(p -> shouldDownload(p, storedPluginsByKey))
+    var downloadSkipReasonByServerPlugin = serverPlugins.stream()
+      .collect(Collectors.toMap(Function.identity(), plugin -> determineIfShouldSkipDownload(plugin, storedPluginsByKey)));
+
+    var pluginsToDownload = downloadSkipReasonByServerPlugin.entrySet().stream()
+      .filter(entry -> entry.getValue().isEmpty())
+      .map(Map.Entry::getKey)
+      .collect(Collectors.toList());
+    var serverPluginsExpectedInStorage = downloadSkipReasonByServerPlugin.entrySet().stream()
+      .filter(entry -> entry.getValue().isEmpty() || entry.getValue().get().equals(DownloadSkipReason.UP_TO_DATE))
+      .map(Map.Entry::getKey)
       .collect(Collectors.toList());
 
     if (pluginsToDownload.isEmpty()) {
       storage.plugins().storeNoPlugins();
+      storage.plugins().cleanUpUnknownPlugins(serverPluginsExpectedInStorage);
       return new PluginSynchronizationSummary(false);
     }
     downloadAll(serverApi, pluginsToDownload, cancelMonitor);
+    storage.plugins().cleanUpUnknownPlugins(serverPluginsExpectedInStorage);
```

Comment:
If I understand this correctly, it is now also cleaning up the old leftover plug-ins like older versions of a analyzer? If yes then this would be neat!

Reply:
Correct. In fact any file that does not correspond to a current plugin on the server will be removed